baremetal/src/include/ebbrt/GeneralPurposeAllocator.h

```diff
@@ -15,6 +15,40 @@
 #include <ebbrt/VMemAllocator.h>
 
 namespace ebbrt {
+
+// page fault handler for mapping in physical pages
+// to virtual pages on all cores
+class LargeRegionFaultHandler : public ebbrt::VMemAllocator::PageFaultHandler {
+  std::vector<ebbrt::Pfn> vecPfns;
+  uintptr_t sAddr;  // start addr of region
+  size_t ps = 2 * 1024 * 1024;  // large page size
+
+ public:
+  void SetAddr(uintptr_t s) { sAddr = s; }
+  void SetVec(std::vector<ebbrt::Pfn> m) { vecPfns = std::move(m); }
+
+  // given faulted address, calculates virtual page frame and finds
+  // corresponding physical page frame in tmap table, then calls
+  // MapMemory to do actual mapping
+  void HandleFault(ebbrt::idt::ExceptionFrame* ef,
+                   uintptr_t faulted_address) override {
+    auto vpage = ebbrt::Pfn::LDown(faulted_address);
+
+    // calculate number of 2Mb pages from sAddr to
+    // use as index to Pfns
+    auto mAddr = vpage.ToLAddr();
+    kassert(mAddr > sAddr);
+    auto index = (mAddr - sAddr) / ps;
+    if (likely(index < vecPfns.size())) {
+      ebbrt::vmem::MapMemoryLarge(mAddr, vecPfns[index], ps);
+    } else {
+      kabort("%s: aborting - cannot find 0x%lx in "
+             "vector.\n",
+             __PRETTY_FUNCTION__, mAddr);
+    }
```

Comment:
if you hit the else case, it's a logic bug, right (e.g. either vmemallocator messed up, or something else? Turn this into `assert(index < vecPfns.size())` if so